### ISSUE TITLE
Bumping to modern tools and architecture - OOP, PSR-4, DI, Symfony components... ready or not?

### ISSUE THREAD

**Issue by TomasVotruba** [squizlabs/PHP_CodeSniffer#1043]
Disclaimer: This just a feedback. I have no intention to criticize. I want to find out, whether our intention meet in some way and cooperate, if possible. 

I have made similar question/request 2 years ago, just wanna check your opinion at the moment. Now I'm much more motivated to do these stuffss.

----

Hi, I have great troubles building wrapper above PHP_CodeSniffer, because it has some architecture drawbacks:

- depending on PHP 5.1 is way deprecated and makes development very hard
    - **it disallows namespaces from PHP 5.3**
    - **it disallows PSR-4**
    - the **only supported versions will be PHP 5.6+** in 15 days
- using constants for configuration makes using or changing any state impossible
- there are lot of God classes, like CLI, CodeSniffer, File, Fixer
- console arguments managements is custom made, instead of using Symfony\Console (PHP version issue) which **would bring more focus on clean code and logic**
- also code quality according to [Scrutinizer is only 5.7](https://scrutinizer-ci.com/g/Symplify/PHP_CodeSniffer/)

It might be caused by the project born date of 2006 (according to Github history).

At the moment, to use this library, I'd have to fork and refactor most of the points above. 
**All these points would make it easier for anyone to contribute or improve the tool.** I just want to ask before I start, **if we have some ideas in common**.

I know you have very strict opinion about using old PHP versions and not to use 3rd party libraries. 

**What do you think about this now?**

**Comment by gsherwood** [comment]
You should look at the 3.0 branch, which I started over a year ago to work on a refactor of the code base. It has to maintain backwards compatibility where possible, but it has allowed me to add features like caching and parallel file processing that are too hard in the current code base.

Having said that, it does use namespaces for all files, including sniff files, so custom sniffs and report classes will not work (I still need to write an upgrade guide for them). But the branch can be tested using the built-in standards and custom rulesets that just use internal sniffs.

As of this post, the current list of high-level changes can be found here (reproduced below): https://github.com/squizlabs/PHP_CodeSniffer/blob/3.0/package.xml#L29-L63
```
- Caching between runs (--cache, --no-cache and "cache" config var) (request #530)
  -- Use --cache=/path/to/cacheFile to use a specific file
- Check files in parallel (--parallel=1 for off, --parallel=100 etc) (request #421)
- Added a new "code" report type to show a code snippet for each error (request #419)
  -- The line containing the error is printed, with 2 lines above and below
  -- The location of the errors is underlined in the code
  -- Use --report=code to try it out
- Added support for custom filtering of the file list
  -- use --filter=/path/to/filter.php to filter the list
  -- extend \PHP_CodeSniffer\Filters\Filter to also get the basic PHPCS ext/path filtering
  -- extend \PHP_CodeSniffer\Filters\ExactMatch to get the basic filtering + ability to use blacklists and whitelists
  -- included filter \PHP_CodeSniffer\Filters\GitModified is a good example of an ExactMatch filter
  -- use --filter=gitmodified to only check modified and added files in a git repo
- Automatic discovery of executable paths (request #571)
  -- Thanks to Sergey Morozov for the patch
- Must pass '-' on the cli to have PHPCS wait for STDIN
  -- Can still pass content on STDIN as normal
- All sniff errors are now exceptions and caught to stop infinite loops
- Sniff codes are no longer optional - use them with addError/Warning/Fixable
- installed_paths config setting can now point directly to a standard instead of the dir above
- --report=full,summary,info instead of --report=full --report=summary --report=info
- Can now set severity, message type and exclude patterns for entire sniff, category or standard
- Can include a single sniff code, which will exclude all other codes from that sniff
  -- if the sniff is already included by an imported standard, set sniff severity to 0 and include the specific message you want
- phpcbf no longer uses patch - files are always overwritten (--no-patch option has been removed)
- Added --basepath option to strip a directory from the front of file paths in output (request #470)
  -- the basepath is absolute or relative to the current directory
  -- to make paths relative to current dir, use --basepath=.
- Ignore rules are now checked when using STDIN (request #733)
- Added include-pattern tag to include a sniff for specific files and folders only (request #656)
  -- exact opposite of the exclude-pattern tag
  -- only useable within sniffs, not globally like exclude-patterns are
- Added -m option to stop error messages from being recorded, which saves a lot of memory
  -- PHPCBF always uses this setting to reduce memory
  -- Setting $recordErrors in custom reports no longer supported
```

> I know you have very strict opinion about using old PHP versions and not to use 3rd party libraries.

I do. The min PHP version in 3.0 is PHP 5.4, but I've had calls to move it back to 5.3. I'm not sure what I'll do yet. Suggesting that 5.6 and 7 are the only supported versions is incorrect. Older versions will continue to be supported by OS vendors for many years to come, and many developers still have to use these older versions in their day-to-day work. It sucks, but that's life, and I have no intention of dropping support for these developers, especially given I've been stuck in these environments as well. So I may very well move back to 5.3, but it really depends on when I get time to do a release and what feedback I get during the alpha and beta stages. I know, for example, that you can officially get PHP 5.4 on REHL6, but that doesn't mean a client will be happy to move to it. But maybe that wont be a problem by the time the 3.0 release comes out.

I'm also very happy with how console management works in PHPCS and I have no intention of ever using a 3rd party library for this. I see absolutely no need for it and I really don't want the tool to have to be "composer installed" just to work. In addition, the assertion that using the Symfony component would bring a focus on clean code and logic is baseless.

I doubt I'd ever accept a PR that brings in a 3rd-party module to replace core code, or that raises the min PHP version, but I'd be very happy for the 3.0 branch to be reviewed so see if there are additional design changes that can be made in there without further breaking backwards compatibility.

**Comment by TomasVotruba** [comment]
I've checked the 3.0 branch briefly and I only see namespaces. Anything else related to my features?

The [PHP 5.6 a 7.0 will be only supported versions](http://php.net/supported-versions.php). 

The common way to maintain older PHP versions for package is to keep 2.x branch open and make hotfixes, while developing in new branch.

**Comment by gsherwood** [comment]
> I've checked the 3.0 branch briefly and I only see namespaces. Anything else related to my features?

I don't have anything written up, so you'll have to take more than a brief look if you want to see how it all works now.

> The PHP 5.6 a 7.0 will be only supported versions.

You completely missed my point. PHP might not be supporting older versions, but OS vendors do. And that's what many many PHP developers around the world, including myself, need to deal with every day.

> The common way to maintain older PHP versions for package is to keep 2.x branch open and make hotfixes, while developing in new branch.

I'm sure some projects do this, but you're not going to ever be able to back up your claim of it being "the common way". But also, I don't care what other projects do. My intention is to ensure all developers, not matter what PHP version you get stuck on, get access to all the new features I can provide. If that makes my life harder because I can't use newer language features (and it does) then that is something I have decided to live with because of my goals.

I'm going to close this because I can't see how this is ever going to be a constructive discussion. I still suggest you look at the 3.0 branch if you want to know what changes I've made to PHPCS, and do not spend time sending PRs for a 2.0 branch refactor as it wont happen (that's what the 3.0 branch is for). But if you'd like to suggest changes to the 3.0 branch, I'd be very happy to review any PRs or comments on it.
